### PR TITLE
Fix fast amg wrapper and generalize RebuildOnUpdatePreconditioner

### PR DESCRIPTION
**Pull request by multitalentloes**
Atgeirr suggested that the famg preconditioner should have an update that reconstructs the preconditioner, this is fixed in this PR.
While doing this the RebuildOnUpdatePreconditioner wrapper class was made more general using parameter packs to keep the factory as clean as possible.

### CODE REVIEW

**Comment by atgeirr** [comment]
jenkins build this please

File: opm/simulators/linalg/PreconditionerWithUpdate.hpp [hunk 1]
```diff
@@ -84,18 +84,49 @@ getDummyUpdateWrapper(Args&&... args)
...
+struct PreconditionerMaker : public GeneralPreconditionerMaker<OriginalPreconditioner> {
+    PreconditionerMaker(Args&&... args)
+        : args_(args...)
+    {
+    }
+    std::unique_ptr<
+        Preconditioner<typename OriginalPreconditioner::domain_type, typename OriginalPreconditioner::range_type>>
+    make() override
+    {
+        return std::unique_ptr<
```
Comment (atgeirr) [inline opm/simulators/linalg/PreconditionerWithUpdate.hpp:110]: I suggest avoiding the explicit `new` by using `make_unique`:
```
return std::apply([](auto...args){
    return std::make_unique<OriginalPreconditioner>(args...);
}, args_);
```

Comment (bska) [reply opm/simulators/linalg/PreconditionerWithUpdate.hpp:110]: > ```c++
> return std::apply(
>     [](auto...args){
>         return std::make_unique<OriginalPreconditioner>(args...);
>     }, args_);
> ```

Better still:
```C++
return std::apply(
    [](auto&&... args) {
        return std::make_unique<OriginalPreconditioner>(std::forward<Args>(args)...);
    }, args_);
```
There's no need to copy the arguments in this case.

Comment (multitalentloes) [reply opm/simulators/linalg/PreconditionerWithUpdate.hpp]: I now use the latest suggestion

**Comment by multitalentloes** [comment]
The latest commits rewrites the wrapper function that is templated with the arguments to the preconditioner constructor such that the wrapper itself does not need to be. This avoids adding extra templates to the factory, while still allowing the RebuildOnUpdatePreconditioner to be generic with respect to construction parameters.

**Review by atgeirr — APPROVED** [review]

**Comment by atgeirr** [comment]
jenkins build this please

**Pull request merged**